pallets/pallet-deposit-storage/src/fungible/mod.rs

```diff
@@ -0,0 +1,238 @@
+where
+	T: Config,
+{
+	/// This function gets called inside all invocations of `hold`, `release`,
+	/// and `release_all`. The value of `amount` is always the final amount that
+	/// should be kept on hold, meaning that a `release_all` will result in a
+	/// value of `0`. An hold increase from `1` to `2` will imply a value of
+	/// `amount` of `2`, i.e., the total amount held at the end of the
+	/// evaluation. Similarly, a decrease from `3` to `1` will imply a `amount`
+	/// value of `1`.
+	fn set_balance_on_hold(reason: &Self::Reason, who: &T::AccountId, amount: Self::Balance) -> DispatchResult {
+		SystemDeposits::<T>::try_mutate_exists(&reason.namespace, &reason.key, |maybe_existing_deposit_entry| {
+			match maybe_existing_deposit_entry {
+				// If this is the first registered deposit for this reason, create a new storage entry.
+				None => {
+					if amount > Zero::zero() {
+						// Increase the held amount for the final runtime hold reason by `amount`.
+						<T::Currency as UnbalancedHold<T::AccountId>>::increase_balance_on_hold(
+							&T::RuntimeHoldReason::from(reason.clone().into()),
+							who,
+							amount,
+							Precision::Exact,
+						)?;
+						*maybe_existing_deposit_entry = Some(DepositEntry {
+							deposit: Deposit {
+								amount,
+								owner: who.clone(),
+							},
+							reason: T::RuntimeHoldReason::from(reason.clone().into()),
+						})
+					}
+					Ok(())
+				}
+				Some(existing_deposit_entry) => {
+					// The pallet assumes each (namespace, key) points to a unique deposit, so the
+					// same combination cannot be used for a different account.
+					if existing_deposit_entry.deposit.owner != *who {
+						return Err(DispatchError::from(Error::<T>::DepositExisting));
+					}
+					if amount.is_zero() {
+						// If trying to remove all holds for this reason (`amount` == 0), decrease the
+						// held amount for the final runtime hold reason by the amount that was held by
+						// this reason.
+						<T::Currency as UnbalancedHold<T::AccountId>>::decrease_balance_on_hold(
+							&T::RuntimeHoldReason::from(reason.clone().into()),
+							who,
+							existing_deposit_entry.deposit.amount,
+							Precision::Exact,
+						)?;
+						// Since we're setting the held amount to `0`, remove the storage entry.
+						*maybe_existing_deposit_entry = None;
+					// We're trying to update (i.e., not creating, not deleting)
+					// the held amount for this hold reason.
+					} else {
+						// If trying to increase the amount held, increase the held amount for the final
+						// runtime hold reason by the (amount - stored amount) difference.
+						if amount >= existing_deposit_entry.deposit.amount {
+							let amount_to_hold = amount
+								.checked_sub(&existing_deposit_entry.deposit.amount)
+								.ok_or_else(|| {
+									log::error!(target: LOG_TARGET, "Failed to evaluate {:?} - {:?}", amount, existing_deposit_entry.deposit.amount);
+									Error::<T>::Internal
+								})?;
```

Comment:
From my point of view, we can also `saturating_sub` here. It is impossible that this operation can fail. We checked the conditions. But it is also fine for me to leave it like that.

Reply:
I mean, the runtime cost of both is the same since they check for underflows, but while with saturating we don't know for sure, with checked we do, so why not using the checked version? We've done that elsewhere as well, in case invariants are broken.

Reply:
It's also good for tests, since the code could change and you could start triggering `silent` saturating stuff, while with this you'd get an explicit `Internal` error, so I don't see any advantages of using saturating.

Reply:
If possible, I would be curious about the case.